runtime/src/utxo.rs

```diff
@@ -144,199 +100,152 @@ decl_module! {
+            
+            total_output = total_output.checked_add(output.value).ok_or("output value overflow")?;
+        }
 
-            let hash = BlakeTwo256::hash_of(output);
-            ensure!(!<UnspentOutputs>::exists(hash), "output already exists");
+        ensure!( total_input >= total_output, "output value must not exceed input value");
 
-            total_output = total_output
-                .checked_add(output.value)
-                .ok_or("output value overflow")?;
+        Ok( total_input.checked_sub(total_output).ok_or("reward underflow")? )
+    }
+    
+    /// Update storage to reflect changes made by transaction
+    /// Where each utxo key is a hash of the entire transaction and its order in the TransactionOutputs vector
+    fn update_storage(transaction: &Transaction, reward: Value) -> DispatchResult {
+        // Calculate new reward total
+        let new_total = <RewardTotal>::get()
+            .checked_add(reward)
+            .ok_or("Reward overflow")?;
+        <RewardTotal>::put(new_total);
+
+        // Storing updated reward value
+        for input in &transaction.inputs {
+            <UtxoStore>::remove(input.outpoint);
         }
 
-        if missing_utxo.is_empty() {
-            ensure!(
-                total_input >= total_output,
-                "output value must not exceed input value"
-            );
-            Ok(CheckInfo::Totals {
-                input: total_input,
-                output: total_output,
-            })
-        } else {
-            Ok(CheckInfo::MissingInputs(missing_utxo))
+        // Add new UTXO to be used by future transactions
+        let index: u64 = 0;
+        for output in &transaction.outputs {
+            let hash = BlakeTwo256::hash_of(&(&transaction.encode(), index));
+            <UtxoStore>::insert(hash, output);
+            index.checked_add(1).ok_or("output index overflow")?;
         }
+
+        Ok(())
     }
-	
-    /// Redistribute combined leftover value evenly among chain authorities
-    fn spend_leftover(authorities: &[H256]) {
-        let leftover = <LeftoverTotal>::take();
-        let share_value: Value = leftover
+
+    /// Redistribute combined reward value evenly among chain authorities
+    fn disperse_reward(authorities: &[H256]) {
+        let reward = <RewardTotal>::take();
+        let share_value: Value = reward
             .checked_div(authorities.len() as Value)
             .ok_or("No authorities")
             .unwrap();
         if share_value == 0 { return }
 
-        let remainder = leftover
+        let remainder = reward
             .checked_sub(share_value * authorities.len() as Value)
             .ok_or("Sub underflow")
             .unwrap();
-        <LeftoverTotal>::put(remainder as Value);
+        <RewardTotal>::put(remainder as Value);
 
         for authority in authorities {
             let utxo = TransactionOutput {
                 value: share_value,
                 pubkey: *authority,
```

Comment:
1. This won't happen because hashes are guaranteed to be unique. `disperse rewards` only executes once during block finalization. so block number will always be different.  each validator only gets one utxo as well

2. we can emit an event or something. validators will also know this, since they are aware of what happens during onfinalize. Will factor this in as a future/later optimization.

I'm not keen on the idea of adding a storage item, since it just behaves like block number. And we only want to store absolutely essential things on chain.

Reply:
You're totally right. I missed that you were already using the block number. And I agree that is better than an additional storage item.

I do still like the idea of an event, and I agree it should happen in a follow-up.